### PR TITLE
Capture changes as unpackaged metadata.

### PR DESCRIPTION
**Pull request by jgerigmeyer**
## Cute animal pic
_Because everyone likes pictures of animals._

goat

## Link to Trello card (if applicable)
_If this PR relates to a Trello card, don't forget to reference this PR on the card as well._

https://trello.com/c/VXUHYQEZ

## Description
_The commit messages say what you did; this should explain why and how._
- [x] Description is in Trello card

## Steps to test/reproduce
_Please explain how to best reproduce the issue and/or test the changes locally._

- `./derrick build`
- `./derrick migrate`
- Create a Dev Org
- Make changes in the org (https://oddbird.slack.com/archives/CH6EN9NMA/p1576536024007000)
- "check again" on Dev Org to find new changes
- Click "Capture Task Changes"

## Impacted areas of website
_Provide URLs and/or a description of where to look._

- Capture-changes modal

## Show me
_Provide screenshots/animated gifs/videos if necessary._

Screenshot 2020-03-04 12 55 51

Screenshot 2020-03-04 12 56 03

Screenshot 2020-03-04 12 56 16

---

## Reviewer tasks:

- [ ] Pulled branch and ran code locally (or viewed on preview deployment, if applicable)
- [ ] Reproduced the issue to review in the browser
- [ ] Reviewed code
- [ ] Updated Trello card (if applicable)

### CODE REVIEW

File: metashare/api/jobs.py [hunk 1]
```diff
@@ -128,6 +131,34 @@ def alert_user_about_expiring_org(*, org, days):
...
 
 
+def _get_valid_target_directories(scratch_org_config):
+    """
+    Expects to be called from within a `local_github_checkout`.
+    """
+    if scratch_org_config.get("source_format") == "sfdx":
+        with open("sfdx-project.json") as f:
+            sfdx_project = json.loads(f.readlines())
+            package_directories = sfdx_project["packageDirectories"]
```
Comment (jgerigmeyer) [inline metashare/api/jobs.py]: @wlonk It looks like you're expecting this to be a list of strings, but according to the docs David linked (https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_ws_config.htm) it seems like it's actually a list of objects. We also might want to track which one is the default?

File: metashare/api/sf_org_changes.py [hunk 2]
```diff
@@ -51,10 +53,10 @@ def run_retrieve_task(user, scratch_org, project_path, desired_changes):
         default_package_directory
```
Comment (jgerigmeyer) [inline metashare/api/sf_org_changes.py]: It seems like the above lines setting `package_directories` and `default_package_directory` can be removed? They're also slightly different than the new code in `jobs._get_valid_target_directories` -- maybe this should be moved over there?

File: metashare/api/views.py [hunk 3]
```diff
@@ -246,7 +246,13 @@ def commit(self, request, pk=None):
             )
         commit_message = serializer.validated_data["commit_message"]
         desired_changes = serializer.validated_data["changes"]
-        scratch_org.queue_commit_changes(request.user, desired_changes, commit_message)
+        target_directory = serializer.validated_data["target_directory"]
```
Comment (jgerigmeyer) [inline metashare/api/views.py]: Do we want to validate that `target_directory` is one of the options in `scratchorg.valid_target_directories`?

File: metashare/api/sf_org_changes.py (hunk 2 above)
Comment (wlonk) [reply metashare/api/sf_org_changes.py]: There's a slight difference in the logic between this code and `get_valid_target_directories`; I'd like to ask @davisagli about it before changing.

David:

`get_valid_target_directories` operates on the assumptions described in https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_ws_config.htm and presumes the first value in `"packageDirectories"` is the default if there is no explicit default. The logic above in _this_ function seeks an explicitly marked default, and depends on that. Should we be tracking the _explicit_ default as opposed to the implicit default? Does that distinction still matter here? Thanks!

Comment (jgerigmeyer) [reply metashare/api/sf_org_changes.py]: @wlonk 

> `get_valid_target_directories` operates on the assumptions described in https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_ws_config.htm and presumes the first value in `"packageDirectories"` is the default if there is no explicit default.

I don't see anything in the docs that say to use the first value:

> If you have only one path, we assume it’s the default, so you don’t have to explicitly set the default parameter. If you have multiple paths, you must indicate which one is the default.

So I assume we should check for an explicit default, or if there's only one value then treat it as the default.

File: metashare/api/sf_org_changes.py [hunk 4]
```diff
@@ -13,7 +13,60 @@
...
 
 
-def run_retrieve_task(user, scratch_org, project_path, desired_changes):
+def get_valid_target_directories(scratch_org_config):
+    """
+    Expects to be called from within a `local_github_checkout`.
+    """
+    package_directories = {}
+    if scratch_org_config.get("source_format") == "sfdx":
+        with open("sfdx-project.json") as f:
```
Comment (jgerigmeyer) [inline metashare/api/sf_org_changes.py]: Do we want a fallback e.g. `if os.path.exists("sfdx-project.json")`?

File: metashare/api/sf_org_changes.py [hunk 5]
```diff
@@ -13,7 +13,60 @@
...
+            sfdx_project = json.loads(f.readlines())
```
Comment (jgerigmeyer) [inline metashare/api/sf_org_changes.py]: Out of curiosity, what's the difference between `json.loads(f.readlines())` and `json.load(f)` used below?

File: metashare/api/models.py [hunk 6]
```diff
@@ -581,6 +581,9 @@ class ScratchOrg(PushMixin, HashIdMixin, TimestampsMixin, models.Model):
     owner_sf_username = StringField(blank=True)
     owner_gh_username = StringField(blank=True)
     has_been_visited = models.BooleanField(default=False)
+    valid_target_directories = JSONField(
+        default=list, encoder=DjangoJSONEncoder, blank=True
+    )
```
Comment (jgerigmeyer) [inline metashare/api/models.py]: It looks like `valid_target_directories` is now a `dict`, not a `list`?

File: metashare/api/views.py [hunk 7]
```diff
@@ -264,7 +264,15 @@ def commit(self, request, pk=None):
...
+        if target_directory not in scratch_org.valid_target_directories:
```
Comment (jgerigmeyer) [inline metashare/api/views.py]: I think `valid_target_directories` is now a `dict`, so doesn't this need to check the nested lists under `valid_target_directories["source"]`, `valid_target_directories["pre"]`, `valid_target_directories["post"]`, and `valid_target_directories["config"]`?

File: metashare/api/sf_org_changes.py (hunk 5 above)
Comment (wlonk) [reply metashare/api/sf_org_changes.py]: They amount to the same thing.

File: metashare/api/sf_org_changes.py (hunk 4 above)
Comment (wlonk) [reply metashare/api/sf_org_changes.py]: Not until we sort out the question above.

File: metashare/api/sf_org_changes.py (hunk 2 above)
Comment (wlonk) [reply metashare/api/sf_org_changes.py]: The docs say "if you have more than one value, you _must_ have one and only one marked as 'default'," so that means that if you have none marked as default, you _must_ have only one, and then we can take the first-and-only as the default. This means that if we have anything the "source" key of the dict that `get_valid_target_directories` returns, the first one will always be the default. I am not sure how this interacts with the logic already here and would like a sanity check on it, because I don't know what assumptions the logic here is built on.

File: metashare/api/sf_org_changes.py (hunk 4 above)
Comment (jgerigmeyer) [reply metashare/api/sf_org_changes.py]: Okay... Even if it is set as `source_format: "sfdx"`, if it does not have a `sfdx-project.json` file then I assume we should fall back to `src` as the default. What else would you propose -- raising an error?

Comment (davisagli) [reply metashare/api/sf_org_changes.py]: That would be an error condition, yeah -- it would just error out when trying to deploy to an org if source_format is sfdx but there's no sfdx-project.json

File: metashare/api/sf_org_changes.py (hunk 2 above)
Comment (jgerigmeyer) [reply metashare/api/sf_org_changes.py]: Understood. I think given the docs we can reasonably assume that the intention is to get the default directory, either explicitly marked or the only (or first, defensively) one.

File: metashare/api/migrations/0059_auto_20200220_1939.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
+# Generated by Django 3.0.3 on 2020-02-20 19:39
```
Comment (jgerigmeyer) [inline metashare/api/migrations/0059_auto_20200220_1939.py]: @wlonk I was going to rename this, but I'm actually going to backport it to `0051` since that's equivalent in terms of causing trouble with your migrations. :)

Comment (wlonk) [reply metashare/api/migrations/0059_auto_20200220_1939.py]: I think you care about the naming and number of migrations more than I do; to me, the cost of changing them once they've been made and pushed is high, and the cost of having many of them and having them have opaque names is low. Would you like me to change my view of them? I am happy to, to try to make things better for you.

Comment (jgerigmeyer) [reply metashare/api/migrations/0059_auto_20200220_1939.py]: I don't feel strongly, but I do think sensible names (not "auto") are a higher priority for me. I don't mind as much having lots of them.

Comment (wlonk) [reply metashare/api/migrations/0059_auto_20200220_1939.py]: Sounds good. I'll prioritize good names.

File: metashare/api/gh.py [hunk 9]
```diff
@@ -114,15 +114,30 @@ def local_github_checkout(user, repo_id, commit_ish=None):
...
+    project_config = get_project_config(**kwargs)
     return project_config.project__git__prefix_feature
 
 
+def get_source_format(**kwargs):
+    """
+    Expects to be in a local_github_checkout.
+    """
+    project_config = get_project_config(**kwargs)
+    return project_config.project__source_format
```
Comment (jgerigmeyer) [inline metashare/api/gh.py]: @wlonk It's not clear to me what the difference is (or which is preferred) between this use of `GlobalConfig` vs the use of `BaseCumulusCI` in `sf_org_changes.run_retrieve_task`. It seems like both end up with the project config (`global_config.get_project_config()` vs `cci.project_config`).

File: metashare/api/sf_org_changes.py [hunk 10]
```diff
@@ -9,16 +9,26 @@
...
 from .sf_run_flow import refresh_access_token
 
 
-def get_valid_target_directories(scratch_org_config):
+def get_valid_target_directories(user, scratch_org, repo_root):
     """
     Expects to be called from within a `local_github_checkout`.
     """
     package_directories = {}
-    if scratch_org_config.get("source_format") == "sfdx":
```
Comment (jgerigmeyer) [inline metashare/api/sf_org_changes.py]: This was not working, because `source_format` is not stored on the `scratch_org_config` -- it's part of the _package_ config from cumulusci.yml (like the branch prefix).

File: metashare/api/sf_org_changes.py [hunk 11]
```diff
@@ -44,21 +54,21 @@ def get_valid_target_directories(scratch_org_config):
...
-            if os.path.isdir(dirname)
+            if os.path.isdir("unpackaged/post/" + dirname)
         ]
     if os.path.isdir("unpackaged/config"):
         package_directories["config"] = [
-            dirname
+            "unpackaged/config/" + dirname
             for dirname in os.listdir("unpackaged/config")
-            if os.path.isdir(dirname)
+            if os.path.isdir("unpackaged/config/" + dirname)
```
Comment (jgerigmeyer) [inline metashare/api/sf_org_changes.py]: @wlonk These changes ensure that we're storing the full path, not just the final dirname. Feel free to cleanup as needed.

File: src/js/components/tasks/capture/directories.tsx [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+  });
+
+  return (
+    <form className="slds-form slds-p-around_large">
+      <div className="slds-form-element__help slds-p-bottom_small">
+        <Trans i18nKey="targetDirectoryInfo">
+          Here’s an explanation of what each type of directory is. The Package
+          Directories are this and the Pre-Install and Post-Install Directories
+          are different. The Optional Configuration Directories are something
+          else entirely.
```
Comment (jgerigmeyer) [inline src/js/components/tasks/capture/directories.tsx]: @davisagli I'd appreciate your suggestions for some explanatory content here.

File: metashare/api/gh.py (hunk 9 above)
Comment (jgerigmeyer) [reply metashare/api/gh.py]: cc @davisagli -- you might have input on the above question.

Comment (wlonk) [reply metashare/api/gh.py]: @davisagli can answer that better than I can.

File: metashare/api/sf_org_changes.py (hunk 11 above)
Comment (wlonk) [reply metashare/api/sf_org_changes.py]: I think they're fine as-is.

**Comment by jgerigmeyer** [comment]
@stacyk This is ready for style cleanup. The main things I'm aware of are:

- The list of changes to capture on the second page of the modal/wizard. It hasn't changed a lot, but we removed "changes" from the right column and it might need some cleanup.

Screenshot 2020-02-27 18 00 12

- The list of changes selected on the third page of the modal/wizard. It's obviously not displaying correctly, and needs some grid cleanup.

Screenshot 2020-02-27 18 01 39

File: metashare/api/gh.py (hunk 9 above)
Comment (davisagli) [reply metashare/api/gh.py]: BaseCumulusCI gives you a "CumulusCI runtime" object which provides access to the global config (representing the contents of cumulusci's own cumulusci.yml), the project config (representing the contents of a specific project's cumulusci.yml), and a keychain (which is meant to be an abstraction for accessing org configs regardless of whether they're in cumulusci.yml or on the filesystem or somewhere else, but MetaShare is only using the keychain as a shortcut to construct an org config that's defined in cumulusci.yml)

GlobalConfig is just the global config, but it also provides `get_project_config` as a shortcut to construct the project config.

The default BaseGlobalConfig and BaseProjectConfig classes defined in CumulusCI (which are the ones that you get if you use BaseCumulusCI) do some things that MetaShare doesn't need or want. Specifically they check for overrides in the user's ~/.cumulusci directory (which is relevant for CLI use but not in a web app) and they use git to determine the project's git URL, branch and commit if not passed explicitly (which MetaShare should generally be able to do). So for MetaShare we created special GlobalConfig and ProjectConfig subclasses in metashare/api/custom_cci_configs.py that avoid that logic.

In order to get the special classes you currently have to construct them directly (or construct GlobalConfig and then use its get_project_config to construct a ProjectConfig). It might make sense to define a MetashareCumulusCIRuntime as a subclass of BaseCumulusCI to provide access to all the configs with an API that's more similar to BaseCumulusCI.

Comment (jgerigmeyer) [reply metashare/api/gh.py]: @davisagli That would make sense to me. Since this PR is working (and not introducing a new way for MS to access package configs), I've created a new story for that cleanup: https://trello.com/c/BtES8zHy

File: metashare/api/sf_org_changes.py [hunk 13]
```diff
@@ -71,7 +126,7 @@ def run_retrieve_task(user, scratch_org, project_path, desired_changes):
     retrieve_components(
         components,
         org_config,
-        path,
+        target_directory,
         md_format,
         extra_package_xml_opts=package_xml_opts,
         namespace_tokenize=False,
```
Comment (jgerigmeyer) [inline metashare/api/sf_org_changes.py]: @davisagli I'm not entirely sure how to functionally test this, but in practice the `target_directory` arg doesn't seem to be making a difference here. I captured changes three times, each into different target directories, but every time the changes are in the `force-app/` directory:

- https://github.com/oddbird/MetaShare-Test/commit/b74d1bfc400b816cc1d11b86da646b1c6353ce6f
- https://github.com/oddbird/MetaShare-Test/commit/97e3cabee04909aa58114be6f00951cd5fc8419c
- https://github.com/oddbird/MetaShare-Test/commit/da4d575e33d197dbe25b39d4eb9e5d3490f41fd4

Maybe I'm looking for the wrong thing?

File: src/js/components/tasks/capture/directories.tsx (hunk 12 above)
Comment (davisagli) [reply src/js/components/tasks/capture/directories.tsx]: @jgerigmeyer gah sorry, I wrote this a few days ago but failed to submit my review.

Package Directories hold the main metadata components of the package you are building. These components will be deployed as an unmanaged package during development, but installed as a managed or unlocked package in production.

Pre-Install Directories are for unmanaged metadata components that should always be deployed prior to deploying or installing the package.

Post-Install Directories are for unmanaged metadata components that should always be deployed after deploying or installing the package.

Optional Configuration Directories are for additional sets of metadata components that will only be deployed on demand.

(Note that some projects may not have all of the types of directories and we should ideally only show the explanations for the ones that exist.)

File: metashare/api/sf_org_changes.py [hunk 14]
```diff
@@ -33,28 +96,20 @@ def run_retrieve_task(user, scratch_org, project_path, desired_changes):
...
-                if package_directory.get("default"):
-                    default_package_directory = package_directory["path"]
     if (
         default_package_directory
         and cci.project_config.project__source_format == "sfdx"
     ):  # pragma: no cover
-        path = os.path.join(project_path, default_package_directory)
         md_format = False
     else:
-        path = os.path.join(project_path, "src")
```
Comment (davisagli) [inline metashare/api/sf_org_changes.py]: @wlonk I think the logic for determining md_format needs to be adjusted. Logic should be: md_format is False if there's an entry in packageDirectories in sfdx-project.json whose path matches target_directory. Otherwise md_format is True.

File: metashare/api/sf_org_changes.py (hunk 13 above)
Comment (davisagli) [reply metashare/api/sf_org_changes.py]: @jgerigmeyer I looked at this a bit and see at least two problems:
1. sf_org_changes.run_retrieve_task is using the wrong logic for determining whether the directory is md_format or not (see this comment: https://github.com/oddbird/MetaShare/pull/191/files/332d480427985d10cbdedf5eaddc85407d10b458#r387230332) -- that leads to incorrectly determining that we're retrieving in sfdx format, which leads to...
2. I'm not sure if it's possible to retrieve components into a non-default sfdx format directory. I need to research if there's something we can pass to the underlying command to make this work; for now the target directory is just getting lost in this scenario and everything ends up in the default package path.

Btw, the .sfdx folder should not be getting captured. It's in .gitignore on MetaShare-Test master but maybe not in your project branch.

Comment (jgerigmeyer) [reply metashare/api/sf_org_changes.py]: @davisagli 

1. I think this is fixed -- take a look if you like.
2. Hm, I'm not 100% sure what "a non-default sfdx format directory" is. Is this something you want us to prevent for now on the MS side of things? Or wait on that until you have a chance to do further research/testing?
3. Actually, it looks like `.sfdx/` is in fact git-ignored on that branch (https://github.com/oddbird/MetaShare-Test/blob/feature/fresh-projects-get-your-projects-here__this-gets-cleared-yes/.gitignore#L2), but it's getting committed anyway. Is that a bug somewhere?

Comment (davisagli) [reply metashare/api/sf_org_changes.py]: 2. It's a directory that is listed in packageDirectories in sfdx-project.json, but not the default package directory. I don't think we are doing this in our projects so it's an edge case we can come back to later.
3. Sounds like a bug. We should not be committing things that are in .gitignore

Comment (jgerigmeyer) [reply metashare/api/sf_org_changes.py]: @davisagli I tried to test this again to look into (3), but even after avoiding/fixing the edge cases being discussed in https://github.com/oddbird/MetaShare/commit/e6ce18131d0c16729a79f883e120d450402e9543#r37616332 and https://github.com/oddbird/MetaShare/commit/e6ce18131d0c16729a79f883e120d450402e9543#r37616357, I'm getting an error when trying to commit to an unpackaged directory:

```
id=no-job-id at=INFO time="2020-03-03 20:47:51.687012" tag=external module=worker msg="default: metashare.api.jobs.commit_changes_from_org(commit_message='Trying: \"unpackaged/pre/opportunity_record_types\"', desired_changes={'CustomField': ['Contact.Favorite_Color__c']}, scratch_org=<ScratchOrg: ScratchOrg object (M137Z1O)>, target_directory='unpackaged/pre/opportunity_record_types', user=<SimpleLazyObject: <User: jgerigmeyer>>) (8435b0f1-7139-4dee-82ef-bf5b5dac3d08)"
id=4b5753793af84117aa76847d5052368f at=INFO time="2020-03-03 20:47:51.693832" tag=external module=logging_middleware method=POST path=/api/scratch-orgs/M137Z1O/commit/ status=202 source_ip=127.0.0.1 user_agent='Mozilla/5.0 (Macintosh;=True Intel=True Mac=True OS=True X=True 10.15;=True rv:73.0)=True Gecko/20100101=True Firefox/73.0'=True time=0.05345654487609863 forwardedfor=None user=n10YoRw
HTTP POST /api/scratch-orgs/M137Z1O/commit/ 202 [0.06, 127.0.0.1:37088]
Traceback (most recent call last):
  File "/app/metashare/api/jobs.py", line 293, in commit_changes_from_org
    commit_changes_to_github(
  File "/app/metashare/api/sf_org_changes.py", line 140, in commit_changes_to_github
    run_retrieve_task(
  File "/app/metashare/api/sf_org_changes.py", line 115, in run_retrieve_task
    retrieve_components(
  File "/usr/local/lib/python3.8/site-packages/cumulusci/tasks/salesforce/sourcetracking.py", line 235, in retrieve_components
    sfdx(
  File "/usr/local/lib/python3.8/site-packages/cumulusci/core/sfdx.py", line 49, in sfdx
    raise Exception(f"Command exited with return code {p.returncode}")
Exception: Command exited with return code 1

id=no-job-id at=ERROR time="2020-03-03 20:48:07.105966" tag=external module=worker msg="Exception: Command exited with return code 1
Traceback (most recent call last):
  File \"/usr/local/lib/python3.8/site-packages/rq/worker.py\", line 884, in perform_job
    rv = job.perform()
  File \"/app/metashare/rq_worker.py\", line 12, in wrapper
    return orig(self)
  File \"/usr/local/lib/python3.8/site-packages/rq/job.py\", line 664, in perform
    self._result = self._execute()
  File \"/usr/local/lib/python3.8/site-packages/rq/job.py\", line 670, in _execute
    return self.func(*self.args, **self.kwargs)
  File \"/app/metashare/api/jobs.py\", line 293, in commit_changes_from_org
    commit_changes_to_github(
  File \"/app/metashare/api/sf_org_changes.py\", line 140, in commit_changes_to_github
    run_retrieve_task(
  File \"/app/metashare/api/sf_org_changes.py\", line 115, in run_retrieve_task
    retrieve_components(
  File \"/usr/local/lib/python3.8/site-packages/cumulusci/tasks/salesforce/sourcetracking.py\", line 235, in retrieve_components
    sfdx(
  File \"/usr/local/lib/python3.8/site-packages/cumulusci/core/sfdx.py\", line 49, in sfdx
    raise Exception(f\"Command exited with return code {p.returncode}\")
Exception: Command exited with return code 1
"
```

Comment (davisagli) [reply metashare/api/sf_org_changes.py]: @jgerigmeyer I'll dig into this tomorrow.

File: metashare/api/sf_org_changes.py [hunk 15]
```diff
@@ -30,39 +94,24 @@ def run_retrieve_task(user, scratch_org, project_path, desired_changes):
...
-    if (
-        default_package_directory
-        and cci.project_config.project__source_format == "sfdx"
-    ):  # pragma: no cover
-        path = os.path.join(project_path, default_package_directory)
-        md_format = False
+    valid_directories, sfdx = get_valid_target_directories(
+        user, scratch_org, project_path
+    )
+    md_format = target_directory not in valid_directories["source"]
```
Comment (jgerigmeyer) [inline metashare/api/sf_org_changes.py]:
```suggestion
    md_format = not (sfdx and target_directory in valid_directories["source"])
```

@wlonk My understanding is that this should also take into account whether or not `sfdx` is truthy?

File: metashare/api/sf_org_changes.py (hunk 13 above)
Comment (jgerigmeyer) [reply metashare/api/sf_org_changes.py]: @davisagli To clarify, I hit that error trying to capture into `unpackaged/pre/*` or `unpackaged/post/*` -- it worked fine capturing into `force-app` (https://github.com/oddbird/MetaShare-Test/commit/d2d8bb2869c51986e8a94e3088e8b6dbed1e5b59).

Comment (davisagli) [reply metashare/api/sf_org_changes.py]: Will be fixed by https://github.com/SFDO-Tooling/CumulusCI/pull/1591 once that makes it into MetaShare.

**Comment by jgerigmeyer** [comment]
@mirisuzanne If you want to review this PR, it's ready.

File: metashare/conftest.py [hunk 16]
```diff
@@ -110,6 +110,7 @@ class Meta:
     task = factory.SubFactory(TaskFactory)
     owner = factory.SubFactory(UserFactory)
     org_type = "Dev"
+    valid_target_directories = {"source": [], "pre": [], "post": [], "config": []}
```
Comment (jgerigmeyer) [inline metashare/conftest.py]:
```suggestion
    valid_target_directories = {"source": []}
```

File: src/js/components/tasks/capture/index.tsx [hunk 17]
```diff
@@ -0,0 +1,261 @@
...
+          ) : (
+            i18n.t('Capture Selected Changes')
+          )
+        }
+        variant="brand"
+        onClick={submitChanges}
+        disabled={capturingChanges || !hasCommitMessage}
+      />,
+    ],
+  ];
```
Comment (jgerigmeyer) [inline src/js/components/tasks/capture/index.tsx]: One list of objects, instead of three lists.

File: src/js/components/tasks/capture/index.tsx [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+          key="page-2-button-1"
+          label={i18n.t('Go Back')}
+          variant="outline-brand"
+          onClick={prevPage}
+        />,
+        // <Button
+        //   key="page-2-button-2"
+        //   label={i18n.t('Ignore Selected Changes')}
+        //   variant="outline-brand"
+        // />,
```
Comment (wlonk) [inline src/js/components/tasks/capture/index.tsx]: I prefer removing things like this entirely to keeping them commented out; is there a compelling reason to leave this commented out?

**Comment by jgerigmeyer** [comment]
@mirisuzanne This is ready for final review. I added screenshots to the PR description, in case you just want to do code review and look at screenshots -- but there are also instructions for reproducing if you want to test it locally.

File: src/js/components/tasks/capture/index.tsx [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+      contents: (
+        <TargetDirectoriesForm
+          key="page-1-contents"
+          directories={directories}
+          inputs={inputs as CommitData}
+          errors={errors}
+          handleInputChange={handleInputChange}
+        />
+      ),
+      footer: (
```
[the remark this answers is not in the document]
Comment (jgerigmeyer) [reply src/js/components/tasks/capture/index.tsx]: Correct -- I don't feel strongly, since the Modal component accepts either type.

File: src/js/components/tasks/capture/index.tsx (hunk 18 above)
Comment (jgerigmeyer) [reply src/js/components/tasks/capture/index.tsx]: Again, I could go either way. I left it in since we anticipate adding it back soon -- but I'm fine removing it to.

File: src/js/components/tasks/capture/changes.tsx [hunk 20]
```diff
@@ -130,83 +87,45 @@ const CaptureModal = ({ orgId, changeset, isOpen, toggleModal }: Props) => {
...
     >
-      <form
-        className="slds-form"
-        data-form="task-capture"
-        onSubmit={submitChanges}
-      >
-        <div className="slds-scrollable_y slds-p-around_large">
-          <div className="form-grid slds-p-around_x-small">
+      <ModalCard>
+        <pre>
```
Comment (mirisuzanne) [inline src/js/components/tasks/capture/changes.tsx]: Surprised that `pre` doesn't create line-break issues. If this is mainly for monospace type, I'd probably just replace the span with a `code` tag?

File: src/js/components/tasks/capture/message.tsx [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+    setExpandedPanels({
+      ...expandedPanels,
+      [groupName]: !expandedPanels[groupName],
+    });
+  };
+
+  return (
+    <form className="slds-form slds-p-around_large">
+      <ModalCard>
+        <pre>
```
Comment (mirisuzanne) [inline src/js/components/tasks/capture/message.tsx]: same, unless there's a particular reason it needs to be a fully pre-formatted block?
